src/openforms/formio/components/custom.py

```diff
@@ -390,31 +391,90 @@ class AddressValueSerializer(serializers.Serializer):
         required=False,
         allow_blank=True,
     )
+    streetName = serializers.CharField(
+        label=_("street name"),
+        help_text=_("Derived street name"),
+        required=False,
+        allow_blank=True,
+    )
+    city = serializers.CharField(
+        label=_("city"),
+        help_text=_("Derived city"),
+        required=False,
+        allow_blank=True,
+    )
+    secretStreetCity = serializers.CharField(
+        label=_("city and street name secret"),
+        help_text=_("Secret for the combination of city and street name"),
+        required=False,
+        allow_blank=True,
+    )
+
+    def __init__(self, **kwargs):
+        self.derive_address = kwargs.pop("derive_address", None)
+        super().__init__(**kwargs)
 
     def validate_postcode(self, value: str) -> str:
         """Normalize the postcode so that it matches the regex from the BRK API."""
         return value.upper().replace(" ", "")
 
+    def validate(self, attrs):
+        attrs = super().validate(attrs)
+
+        city = attrs.get("city", "")
+        street_name = attrs.get("streetName", "")
+
+        if self.derive_address:
+            existing_hmac = attrs.get("secretStreetCity", "")
+            postcode = attrs.get("postcode", "")
+            number = attrs.get("houseNumber", "")
+
+            computed_hmac = salt_location_message(
+                {
+                    "postcode": postcode,
+                    "number": number,
+                    "city": city,
+                    "street_name": street_name,
+                }
+            )
+
+            if not constant_time_compare(existing_hmac, computed_hmac):
+                raise serializers.ValidationError(
+                    _("Invalid secret city - street name combination"),
+                    code="invalid",
+                )
+
+        return attrs
+
 
 @register("addressNL")
-class AddressNL(BasePlugin):
+class AddressNL(BasePlugin[AddressNLComponent]):
 
     formatter = AddressNLFormatter
 
-    def build_serializer_field(self, component: Component) -> AddressValueSerializer:
+    def build_serializer_field(
+        self, component: AddressNLComponent
+    ) -> AddressValueSerializer:
 
         validate = component.get("validate", {})
         required = validate.get("required", False)
 
         extra = {}
         validators = []
+        kwargs = {}
         if plugin_ids := validate.get("plugins", []):
             validators.append(PluginValidator(plugin_ids))
 
         extra["validators"] = validators
 
+        if derive_address := component.get("deriveAddress") is not None:
+            kwargs["derive_address"] = derive_address
+
         return AddressValueSerializer(
-            required=required, allow_null=not required, **extra
+            required=required,
+            allow_null=not required,
+            **extra,
+            **kwargs,
         )
```

Comment:
Since we have the migration + import conversion to add this key, we can rely on it and simply pass the kwarg directly:

```py
        return AddressValueSerializer(
            derive_address=component["deriveAddress"],
            required=required,
            allow_null=not required,
            **extra,
            **kwargs,
        )
```